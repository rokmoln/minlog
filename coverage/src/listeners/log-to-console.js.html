<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/listeners/log-to-console.js</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> / <a href="index.html">src/listeners</a> log-to-console.js
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">23.73% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>14/59</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">32.43% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>12/37</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">25% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>1/4</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">23.73% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>14/59</span>
      </div>
    </div>
  </div>
  <div class='status-line low'></div>
<pre><table class="coverage">
<tr><td class="line-count quiet">1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146</td><td class="line-coverage quiet"><span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">1x</span>
<span class="cline-any cline-yes">70x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">28x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">70x</span>
<span class="cline-any cline-yes">20x</span>
<span class="cline-any cline-yes">50x</span>
<span class="cline-any cline-yes">12x</span>
<span class="cline-any cline-yes">38x</span>
<span class="cline-any cline-yes">8x</span>
<span class="cline-any cline-yes">30x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">11x</span>
<span class="cline-any cline-yes">19x</span>
<span class="cline-any cline-yes">2x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-yes">17x</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-no">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span>
<span class="cline-any cline-neutral">&nbsp;</span></td><td class="text"><pre class="prettyprint lang-js">import _ from 'lodash';
import moment from 'moment';
&nbsp;
export let _levelToConsoleFun = function({level, levels}) {
  if (_.isString(level)) {
    // eslint-disable-next-line prefer-destructuring
    level = levels[level];
  }
&nbsp;
  if (_.inRange(level, 0, levels.warn)) {
    return 'error';
  } else if (_.inRange(level, levels.warn, levels.info)) {
    return 'warn';
  } else if (_.inRange(level, levels.info, levels.debug)) {
    return 'info';
  } else if (_.inRange(level, levels.debug, levels.trace)) {
    // return 'debug';
    // console.debug doesn't seem to print anything,
    // but console.debug is an alias to console.log anyway
    return 'log';
  } else if (level === levels.trace) {
    return 'trace';
  }
&nbsp;
  return 'log';
};
&nbsp;
/*
cfg has 2 properties
- level (optional, defaults to trace)
  Any log entry less important that cfg.level is ignore.
- iframeId (optional, default to 'top' or '?'
  An identifier for the current "window".
*/
&nbsp;
export default <span class="fstat-no" title="function not covered" >fu</span>nction(cfg = <span class="branch-0 cbranch-no" title="branch not covered" >{})</span> {
<span class="cstat-no" title="statement not covered" >  return <span class="fstat-no" title="function not covered" >as</span>ync function({entry, logger, rawEntry}) {</span>
<span class="cstat-no" title="statement not covered" >    if (_.filter(rawEntry._args).length === 1 &amp;&amp; rawEntry._args[0]._babelSrc) {</span>
<span class="cstat-no" title="statement not covered" >      return;</span>
    }
&nbsp;
    let maxLevelName = <span class="cstat-no" title="statement not covered" >cfg.level || 'trace';</span>
    let maxLevel = <span class="cstat-no" title="statement not covered" >logger.levels[maxLevelName];</span>
<span class="cstat-no" title="statement not covered" >    maxLevel = _.floor(maxLevel / 10) * 10 + 10 - 1; </span>// round up to next level, not inclusive
<span class="cstat-no" title="statement not covered" >    if (entry.level &gt; maxLevel) {</span>
<span class="cstat-no" title="statement not covered" >      return;</span>
    }
&nbsp;
    let now = <span class="cstat-no" title="statement not covered" >moment(entry._time.stamp).utcOffset(entry._time.utc_offset).toISOString();</span>
    let levelName = <span class="cstat-no" title="statement not covered" >logger.levelToLevelName(entry._level);</span>
    let formattedLevelName = <span class="cstat-no" title="statement not covered" >_.padStart(_.toUpper(levelName), '5');</span>
    let consoleFun = <span class="cstat-no" title="statement not covered" >exports._levelToConsoleFun({</span>
      level: entry._level,
      levels: logger.levels
    });
&nbsp;
    let color = <span class="cstat-no" title="statement not covered" >'';</span>
<span class="cstat-no" title="statement not covered" >    switch (consoleFun) {</span>
    case 'log':
    case 'info':
    case 'trace':
<span class="cstat-no" title="statement not covered" >      color = 'color: dodgerblue';</span>
<span class="cstat-no" title="statement not covered" >      break;</span>
    default:
    }
&nbsp;
    let prefixFormat = <span class="cstat-no" title="statement not covered" >'%c%s %c%s%c';</span>
    let prefixArgs = <span class="cstat-no" title="statement not covered" >[</span>
      color,
      now,
      'font-weight: bold',
      formattedLevelName,
      color
    ];
&nbsp;
    let src = <span class="cstat-no" title="statement not covered" >'';</span>
<span class="cstat-no" title="statement not covered" >    if (entry._babelSrc) {</span>
<span class="cstat-no" title="statement not covered" >      src = _.merge({}, entry._src, entry._babelSrc);</span>
<span class="cstat-no" title="statement not covered" >      src = ` @webpack:///./${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;</span>
    } else <span class="cstat-no" title="statement not covered" >if (entry._src) {</span>
<span class="cstat-no" title="statement not covered" >      src = entry._src;</span>
<span class="cstat-no" title="statement not covered" >      src = ` ${src.file}:${src.line}:${src.column}${src.function ? ` in ${src.function}()` : ''}`;</span>
    }
&nbsp;
    let iframeId = <span class="cstat-no" title="statement not covered" >_.defaultTo(</span>
      cfg.iframeId,
      window.parent === window ? 'top' : '?'
    );
&nbsp;
    let context = <span class="cstat-no" title="statement not covered" >{</span>
      window,
      documentElement: window.document.documentElement
    };
&nbsp;
    let srcFormat = <span class="cstat-no" title="statement not covered" >'%s in the %s context';</span>
    let srcArgs = <span class="cstat-no" title="statement not covered" >[</span>
      src,
      iframeId
    ];
&nbsp;
    let msgFormat = <span class="cstat-no" title="statement not covered" >'';</span>
    let msgArgs = <span class="cstat-no" title="statement not covered" >[];</span>
<span class="cstat-no" title="statement not covered" >    if (entry.msg) {</span>
<span class="cstat-no" title="statement not covered" >      msgFormat = '\n%s';</span>
<span class="cstat-no" title="statement not covered" >      msgArgs = [</span>
        entry.msg
      ];
    }
&nbsp;
    let extraFormat = <span class="cstat-no" title="statement not covered" >'';</span>
    let extraArgs = <span class="cstat-no" title="statement not covered" >[];</span>
&nbsp;
    let extra = <span class="cstat-no" title="statement not covered" >_.omit(rawEntry, [</span>
      '_args',
      '_babelSrc',
      '_level',
      '_src',
      '_time',
      'iframeId',
      'msg'
    ]);
<span class="cstat-no" title="statement not covered" >    _.merge(extra, context);</span>
&nbsp;
    // devTools console sorts keys when object is expanded
<span class="cstat-no" title="statement not covered" >    extra = _.toPairs(extra);</span>
<span class="cstat-no" title="statement not covered" >    extra = _.sortBy(extra, 0);</span>
<span class="cstat-no" title="statement not covered" >    extra = _.fromPairs(extra);</span>
&nbsp;
    // devTools collapses objects with 'too many' keys,
    // so we output objects with only one key
<span class="cstat-no" title="statement not covered" >    _.forEach(extra, <span class="fstat-no" title="function not covered" >fu</span>nction(value, key) {</span>
<span class="cstat-no" title="statement not covered" >      extraArgs.push('\n');</span>
<span class="cstat-no" title="statement not covered" >      extraArgs.push({[key]: value});</span>
    });
&nbsp;
    // eslint-disable-next-line no-console
<span class="cstat-no" title="statement not covered" >    console[consoleFun](</span>
      `${prefixFormat}${srcFormat}:${msgFormat}${extraFormat}`,
      ...prefixArgs,
      ...srcArgs,
      ...msgArgs,
      ...extraArgs
    );
  };
}
&nbsp;</pre></td></tr>
</table></pre>
<div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Sep 24 2017 16:59:24 GMT+0200 (CEST)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
</body>
</html>
