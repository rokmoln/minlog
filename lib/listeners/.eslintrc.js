"use strict";module.exports = {
  extends: [
  'firecloud/browser.js'] };

//# sourceMappingURL=.eslintrc.js.map