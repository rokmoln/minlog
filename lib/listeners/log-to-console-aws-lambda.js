"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = exports.logToConsoleAwsLambda = exports.format = exports._isAwsLambda = exports._nonBreakingWhitespace = void 0;var _lodashFirecloud = _interopRequireDefault(require("lodash-firecloud"));
var _fastSafeStringify = _interopRequireDefault(require("fast-safe-stringify"));







var _util = require("../util");




var _logToConsole = require("./log-to-console");function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}


















let _nonBreakingWhitespace = 'Â ';exports._nonBreakingWhitespace = _nonBreakingWhitespace;

let _isAwsLambda = function () {
  let isNode = typeof process !== 'undefined' && _lodashFirecloud.default.isDefined(_lodashFirecloud.default.get(process, 'versions.node'));
  if (!isNode) {
    return false;
  }

  let isAwsLambda = _lodashFirecloud.default.isDefined(process.env.LAMBDA_TASK_ROOT);
  if (!isAwsLambda) {
    return false;
  }

  return true;
};exports._isAwsLambda = _isAwsLambda;

let format = function (...formatArgs) {
  let [
  format,
  ...args] =
  formatArgs;

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let chunk = require('util').format(format, ...args);
  chunk = _lodashFirecloud.default.replace(chunk, /\n/g, '\r');
  chunk = `${chunk}\n`;
  process.stdout.write(chunk);
};exports.format = format;

let logToConsoleAwsLambda = function (cfg = {}) {
  if (!exports._isAwsLambda()) {
    // use vanilla logger e.g. behind aws-lambda-proxy
    return (0, _logToConsole.logToConsole)(cfg);
  }

  // from https://github.com/Financial-Times/lambda-logger
  // This does make process.stdout.write a blocking function (process.stdout._handle.setBlocking(true);),
  // as AWS Lambda previously streamed to an output which was synchronous,
  // but has since changed to asynchronous behaviour, leading to lost logs.
  if (_lodashFirecloud.default.isFunction(_lodashFirecloud.default.get(process, 'stdout._handle.setBlocking'))) {
    // @ts-ignore
    process.stdout._handle.setBlocking(true);
  }

  // eslint-disable-next-line complexity
  return async function ({ entry, logger, rawEntry }) {
    cfg = _lodashFirecloud.default.isFunction(cfg) ? await (async createError => {try {return await cfg();} catch (_awaitTraceErr) {let err = createError();_awaitTraceErr.stack += "\n...\n" + err.stack;throw _awaitTraceErr;}})(() => new Error()) : await (async createError => {try {return await cfg;} catch (_awaitTraceErr2) {let err = createError();_awaitTraceErr2.stack += "\n...\n" + err.stack;throw _awaitTraceErr2;}})(() => new Error());

    if (_lodashFirecloud.default.isDefined(rawEntry) &&
    _lodashFirecloud.default.filter(rawEntry._args).length === 1 &&
    _lodashFirecloud.default.isDefined(rawEntry._args[0]._babelSrc)) {
      return;
    }

    if (_lodashFirecloud.default.isDefined(cfg.level) && logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      now,
      formattedLevelName,
      src,
      msg
      // @ts-ignore
    } = (0, _logToConsole.serialize)({ entry, logger, rawEntry, cfg });
    cfg = cfg2;

    // prefer JSON output over util.inspect output
    let extra = _lodashFirecloud.default.omit(entry, [
    '_args']);

    let extraStr = (0, _fastSafeStringify.default)(extra, _util.jsonStringifyReplacer, 2);
    extraStr = `\n${extraStr}`;

    // maintain whitespace (looking at you AWS CloudWatch WebUI)
    // by replacing space with non-breaking space
    extraStr = _lodashFirecloud.default.replace(extraStr, / /g, exports._nonBreakingWhitespace);

    let extraArgs = [
    extraStr];


    let formatPairs = [];

    // timestamp
    formatPairs.push([
    '%s',
    now]);


    // awsRequestId
    formatPairs.push([
    '\t%s',
    _lodashFirecloud.default.get(entry, 'ctx.awsRequestId', '-')]);


    // level name
    formatPairs.push([
    '\t%s',
    formattedLevelName]);


    // src
    if (_lodashFirecloud.default.isDefined(src)) {
      formatPairs.push([
      '\t%s',
      src]);

    }

    // msg
    // eslint-disable-next-line lodash/prop-shorthand
    let argNames = _lodashFirecloud.default.keys((0, _util.keepOnlyExtra)(entry));
    if (!_lodashFirecloud.default.isEmpty(argNames)) {
      msg = `${msg} (${_lodashFirecloud.default.join(argNames, ', ')})`;
    }
    msg = `${msg} `;
    msg = _lodashFirecloud.default.padEnd(msg, 255, exports._nonBreakingWhitespace);
    msg = `${msg}.`;
    formatPairs.push([
    '\n%s',
    msg]);


    let formatArgs = (0, _logToConsole.toFormatArgs)(...formatPairs);
    formatArgs.push(...extraArgs);

    exports.format(...formatArgs);
  };
};exports.logToConsoleAwsLambda = logToConsoleAwsLambda;var _default = exports.logToConsoleAwsLambda;exports.default = _default;

//# sourceMappingURL=log-to-console-aws-lambda.js.map