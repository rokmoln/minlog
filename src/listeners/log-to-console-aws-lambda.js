import _ from 'lodash-firecloud';
import fastSafeStringify from 'fast-safe-stringify';
import util from 'util';

import {
  logToConsole,
  serialize as serializeLogToConsole,
  toFormatArgs
} from './log-to-console';

let _nonBreakingWhitespace = 'Â ';

let _isBrowser = typeof window !== 'undefined';
let _isNode = typeof process !== 'undefined' && _.isDefined(_.get(process, 'versions.node'));
let _isAwsLambda = _isNode && _.isDefined(process.env.LAMBDA_TASK_ROOT);

/*
cfg has 1 property
- level (optional, defaults to trace)
  Any log entry less important that cfg.level is ignored.
*/

export let logToConsoleAwsLambda = function(cfg = {}) {
  if (!_isAwsLambda) {
    // use vanilla logger e.g. behind aws-lambda-proxy
    return logToConsole(cfg);
  }

  // from https://github.com/Financial-Times/lambda-logger
  // This does make process.stdout.write a blocking function (process.stdout._handle.setBlocking(true);),
  // as AWS Lambda previously streamed to an output which was synchronous,
  // but has since changed to asynchronous behaviour, leading to lost logs.
  if (_.isFunction(_.get(process, 'stdout._handle.setBlocking'))) {
    process.stdout._handle.setBlocking(true);
  }

  // eslint-disable-next-line complexity
  return async function({entry, logger, rawEntry}) {
    if (_.isDefined(rawEntry) &&
        _.filter(rawEntry._args).length === 1 &&
        rawEntry._args[0]._babelSrc) {
      return;
    }

    if (_.isDefined(cfg.level) && logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      now,
      formattedLevelName,
      src,
      msg,
      extra
    } = serializeLogToConsole({entry, logger, rawEntry, cfg});
    cfg = cfg2;

    msg = `${msg} `;
    msg = _.padEnd(msg, 255, _nonBreakingWhitespace);
    msg = `${msg}.`;

    // prefer JSON output over util.inspect output
    extra = fastSafeStringify(extra, undefined, 2);
    extra = `\n${extra}`;

    // maintain whitespace (looking at you AWS CloudWatch WebUI)
    // by replacing space with non-breaking space
    extra = _.replace(extra, / /g, _nonBreakingWhitespace);

    let extraArgs = [
      extra
    ];

    let formatArgs = [];

    // timestamp
    formatArgs.push([
      '%s',
      now
    ]);

    // awsRequestId
    formatArgs.push([
      ' %s',
      // skip for readability, still available in 'extra'
      // entry.context.awsRequestId
      '-'
    ]);

    // level name
    formatArgs.push([
      '\t%s',
      formattedLevelName
    ]);

    // src
    if (src) {
      formatArgs.push([
        ' %s',
        src
      ]);
    }

    // msg
    formatArgs.push([
      '\n%s',
      msg
    ]);

    formatArgs = toFormatArgs(...formatArgs);
    formatArgs = _.concat(formatArgs, extraArgs);

    // eslint-disable-next-line global-require
    let chunk = util.format(...formatArgs);
    chunk = _.replace(chunk, /\n/g, '\r');
    chunk = `${chunk}\n`;
    process.stdout.write(chunk);
  };
};

export default logToConsoleAwsLambda;
