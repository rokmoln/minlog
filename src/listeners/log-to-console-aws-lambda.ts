import _ from 'lodash-firecloud';
import fastSafeStringify from 'fast-safe-stringify';

import {
  MinLogFormatArgs,
  MinLogLevel,
  MinLogListener
} from '../types';

import {
  jsonStringifyReplacer,
  keepOnlyExtra
} from '../util';

import {
  logToConsole,
  serialize as serializeLogToConsole,
  toFormatArgs
} from './log-to-console';

import {
  Fn,
  MaybePromise
} from 'lodash-firecloud/types';

export interface Cfg {

  /**
   * Any log entry less important that cfg.level is ignored.
   */
  level?: MinLogLevel;
}

let _nonBreakingWhitespace = 'Â ';

let _isAwsLambda = function(): boolean {
  let isNode = typeof process !== 'undefined' && _.isDefined(_.get(process, 'versions.node'));
  if (!isNode) {
    return false;
  }

  let isAwsLambda = _.isDefined(process.env.LAMBDA_TASK_ROOT);
  if (!isAwsLambda) {
    return false;
  }

  return true;
};

export let format = function(...formatArgs: MinLogFormatArgs): void {
  let [
    format,
    ...args
  ] = formatArgs;

  // eslint-disable-next-line @typescript-eslint/no-require-imports
  let chunk = require('util').format(format, ...args);
  chunk = _.replace(chunk, /\n/g, '\r');
  chunk = `${chunk}\n`;
  process.stdout.write(chunk);
};

export let logToConsoleAwsLambda = function(cfg: MaybePromise<Cfg> | Fn<MaybePromise<Cfg>> = {}): MinLogListener {
  if (!_isAwsLambda()) {
    // use vanilla logger e.g. behind aws-lambda-proxy
    return logToConsole(cfg);
  }

  // from https://github.com/Financial-Times/lambda-logger
  // This does make process.stdout.write a blocking function (process.stdout._handle.setBlocking(true);),
  // as AWS Lambda previously streamed to an output which was synchronous,
  // but has since changed to asynchronous behaviour, leading to lost logs.
  if (_.isFunction(_.get(process, 'stdout._handle.setBlocking'))) {
    // @ts-ignore
    process.stdout._handle.setBlocking(true);
  }

  // eslint-disable-next-line complexity
  return async function({entry, logger, rawEntry}) {
    cfg = _.isFunction(cfg) ? await cfg() : await cfg;

    // handle https://github.com/tobiipro/babel-preset-firecloud#babel-plugin-firecloud-src-arg-default-config-needed
    if (_.isDefined(rawEntry) &&
        _.filter(rawEntry._args).length === 1 &&
        _.isDefined(rawEntry._args[0]._babelSrc)) {
      return;
    }

    if (_.isDefined(cfg.level) && logger.levelIsBeyondGroup(entry._level, cfg.level)) {
      return;
    }

    let {
      cfg: cfg2,
      now,
      formattedLevelName,
      src,
      msg
    // @ts-ignore
    } = serializeLogToConsole({entry, logger, rawEntry, cfg});
    cfg = cfg2;

    // use relative path, shorter output
    src = _.replace(src, /\/var\/task\//g, './');

    // prefer JSON output over util.inspect output
    let extra = _.omit(entry, [
      '_args'
    ]);
    let extraStr = fastSafeStringify(extra, jsonStringifyReplacer, 2);
    extraStr = `\n${extraStr}`;

    // maintain whitespace (looking at you AWS CloudWatch WebUI)
    // by replacing space with non-breaking space
    extraStr = _.replace(extraStr, / /g, _nonBreakingWhitespace);

    let extraArgs = [
      extraStr
    ];

    let formatPairs = [];

    // timestamp
    formatPairs.push([
      '%s',
      now
    ]);

    // awsRequestId
    formatPairs.push([
      '\t%s',
      _.get(entry, 'ctx.awsRequestId', '-')
    ]);

    // level name
    formatPairs.push([
      '\t%s',
      formattedLevelName
    ]);

    // src
    if (_.isDefined(src)) {
      formatPairs.push([
        '\t%s',
        src
      ]);
    }

    // msg
    // eslint-disable-next-line lodash/prop-shorthand
    let argNames = _.keys(keepOnlyExtra(entry));
    if (!_.isEmpty(argNames)) {
      msg = `${msg} (${_.join(argNames, ', ')})`;
    }
    msg = `${msg} `;
    msg = _.padEnd(msg, 255, _nonBreakingWhitespace);
    msg = `${msg}.`;
    formatPairs.push([
      '\n%s',
      msg
    ]);

    let formatArgs = toFormatArgs(...formatPairs);
    formatArgs.push(...extraArgs);

    format(...formatArgs);
  };
};

export default logToConsoleAwsLambda;
