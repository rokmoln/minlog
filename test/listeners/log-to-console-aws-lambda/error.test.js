import * as jestDateMock from 'jest-date-mock';
import * as logToConsoleAwsLambdaModule from '../../../src/listeners/log-to-console-aws-lambda';
import _ from 'lodash-firecloud';

import {
  logger
} from './logger';

let _nonBreakingWhitespace = 'Â ';

describe('logToConsoleAwsLambda listener', function() {
  beforeEach(function() {
    jestDateMock.advanceTo(0);
  });

  afterEach(function() {
    jestDateMock.clear();
  });

  it('should print an error', async function() {
    let loggerCallArgs = [
      new Error('This is an error.')
    ];
    // keep only one-level of the stacktrace
    loggerCallArgs[0].stack = _.join([
      _.split(loggerCallArgs[0].stack, '\n')[0]
    ], '\n');

    let snapshot;
    let spyFormat = jest.spyOn(logToConsoleAwsLambdaModule, 'format');

    spyFormat.mockImplementationOnce(function(format, ...formatArgs) {
      snapshot = {
        format,
        formatArgs
      };

      let [
        _now,
        _hyphen,
        _level,
        _src,
        msg,
        objectStr
      ] = formatArgs;

      // JSON.parse cannot handle non-breaking whitespace
      objectStr = _.replace(objectStr, new RegExp(_nonBreakingWhitespace, 'g'), ' ');

      let cond1 = _.startsWith(msg, loggerCallArgs[0].toString());

      let object = JSON.parse(objectStr);
      let cond2 = _.isMatch(object, {
        err: {
          name: loggerCallArgs[0].name,
          message: loggerCallArgs[0].message
        }
      });

      let cond = cond1 && cond2;
      expect(cond).toBe(true);
    });

    await logger.error(...loggerCallArgs);
    expect(spyFormat).toHaveBeenCalledTimes(1);
    expect(snapshot).toMatchSnapshot();
  });
});
