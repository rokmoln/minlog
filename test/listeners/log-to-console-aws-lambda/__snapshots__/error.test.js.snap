// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logToConsoleAwsLambda listener should print an error 1`] = `
Object {
  "format": "%s	%s	%s	%s
%s",
  "formatArgs": Array [
    "1970-01-01T00:00:00.000Z",
    "-",
    "ERROR",
    "",
    "Error: This is an error.                                                                                                                                                                                                                                       .",
    "
{
  \\"_time\\": {
    \\"stamp\\": \\"1970-01-01T00:00:00.000Z\\",
    \\"localStamp\\": \\"1970-01-01T01:00:00.000+01:00\\",
    \\"zone\\": \\"Europe/Stockholm\\",
    \\"utc_offset\\": 60
  },
  \\"_level\\": 30,
  \\"err\\": {
    \\"name\\": \\"Error\\",
    \\"message\\": \\"This is an error.\\",
    \\"stack\\": [
      \\"Error: This is an error.\\"
    ]
  },
  \\"msg\\": \\"Error: This is an error.\\"
}",
  ],
}
`;
